File: topcoder/2019 Topcoder Open Algo_Div_/EllysTeleport.java
import java.io.*;
import java.util.*;

public class EllysTeleport {
	public int getMax(int N, int H1, int A, int B, int P, int Q, int M) {
		return 0;
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("EllysTeleport (500 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            int N = Integer.parseInt(Reader.nextLine());
            int H1 = Integer.parseInt(Reader.nextLine());
            int A = Integer.parseInt(Reader.nextLine());
            int B = Integer.parseInt(Reader.nextLine());
            int P = Integer.parseInt(Reader.nextLine());
            int Q = Integer.parseInt(Reader.nextLine());
            int M = Integer.parseInt(Reader.nextLine());
            Reader.nextLine();
            int __answer = Integer.parseInt(Reader.nextLine());

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(N, H1, A, B, P, Q, M, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1556708543;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(int N, int H1, int A, int B, int P, int Q, int M, int __expected) {
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		EllysTeleport instance = new EllysTeleport();
		int __result = 0;
		try {
			__result = instance.getMax(N, H1, A, B, P, Q, M);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__result == __expected) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + __expected);
			System.err.println("           Received: " + __result);
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "EllysTeleport.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}

File: topcoder/2019 Topcoder Open Algo_Div_/EllysTeleport.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <stack>
#include <map>
#include <algorithm>
#include <set>
#include <numeric>
#include <bitset>
#include <sstream>
#include <utility>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cmath>
#include <cctype>
#include <climits>
#include <fstream>
static const int MOD = 1000000007;

using u32 = unsigned;
using namespace std;

using ll = long long;
static const ll INF = (ll)MOD*MOD;

class EllysTeleport {
    public:
    int getMax(int N, int H1, int A, int B, int P, int Q, int M) {
        vector<ll> v(N);
        v[0] = H1;
        for (int i = 1; i < N; ++i) {
            v[i] = (v[i-1]*A+B)%M;
        }
        sort(v.begin(),v.end());
        vector<ll> to(N);
        for (int i = 0; i < N; ++i) {
            ll t = (v[i]*P+Q) % M;
            if(t < v[0]) to[i] = -1;
            else {
                ll p = upper_bound(v.begin(),v.end(), t)-v.begin()-1;
                to[i] = p;
            }
        }
        int ans = 0;
        for (int i = 0; i < N; ++i) {
            vector<int> visited(N);
            int cur = i, move = 0;
            while(!visited[cur]){
                visited[cur] = 1;
                move++;
                if(to[cur] == -1) break;
                cur = to[cur];
            }
            ans = max(ans, move);
        }
        return ans;
    }
};

// CUT begin
ifstream data("EllysTeleport.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(int N, int H1, int A, int B, int P, int Q, int M, int __expected) {
    time_t startClock = clock();
    EllysTeleport *instance = new EllysTeleport();
    int __result = instance->getMax(N, H1, A, B, P, Q, M);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        int N;
        from_stream(N);
        int H1;
        from_stream(H1);
        int A;
        from_stream(A);
        int B;
        from_stream(B);
        int P;
        from_stream(P);
        int Q;
        from_stream(Q);
        int M;
        from_stream(M);
        next_line();
        int __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(N, H1, A, B, P, Q, M, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1556708627;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "EllysTeleport (500 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end
